readthedocs/redirects/querysets.py

```diff
@@ -25,7 +30,56 @@ def api(self, user=None, detail=True):
         return queryset
 
     def get_redirect_path_with_status(self, path, full_path=None, language=None, version_slug=None):
-        for redirect in self.select_related('project'):
+        # add extra fields with the ``path`` and ``full_path`` to perform a
+        # filter at db level instead with Python
+        queryset = self.annotate(
+            path=Value(
+                path,
+                output_field=CharField(),
+            ),
+            full_path=Value(
+                full_path,
+                output_field=CharField(),
+            ),
+        )
+        prefix = Q(
+            redirect_type='prefix',
+            path__startswith=F('from_url'),
+        )
+        page = Q(
+            redirect_type='page',
+            path__exact=F('from_url'),
+        )
+        exact = (
+            Q(
+                redirect_type='exact',
+                from_url__endswith='$rest',
+                full_path__startswith=F('from_url_without_rest'),
+            ) | Q(
+                redirect_type='exact',
+                full_path__exact=F('from_url'),
+            )
+        )
+        sphinx_html = (
+            Q(
+                redirect_type='sphinx_html',
+                path__endswith='/',
+            ) | Q(
+                redirect_type='sphinx_html',
+                path__endswith='/index.html',
+            )
+        )
+        sphinx_htmldir = Q(
+            redirect_type='sphinx_htmldir',
+            path__endswith='.html',
+        )
+
+        queryset = queryset.filter(prefix | page | exact | sphinx_html | sphinx_htmldir)
+
+        # There should be one and only one redirect returned by this query. I
+        # can't think in a case where there can be more at this point. I'm
```

Comment:
A user could create several redirects with the same value, or with `$rest` in different positions, like `/en/latest/foo/$rest` and `/en/latest/foo/bar/$rest`

Reply:
Probably we should sort the result by creation date if we don't do that already

Reply:
> like `/en/latest/foo/$rest` and `/en/latest/foo/bar/$rest`

Good point. Although, the first one contains the second one in this case, so it does not make too much sense at this point.

We could implement _new feature_ that the ones that matches _more_ gets priority, though. But that's a different story.